asa/asa.1
.Id $Id: asa.1,v 1.1 1998/03/07 00:15:42 SOFTWAY_COM+donn Exp $
.TH ASA 1 "April 1994" POSIX.2 "Linux Programmer's Manual"
.SH NAME
asa \- convert ASA (Fortran) carriage control charachters
.SH SYNOPSIS
.B asa
.RI [ file ...]
.SH DESCRIPTION
.B asa
reads text with ASA cariage control characters from the files specified
as its command arguments, or standard input if none are specified, and
converts them into line printer control sequences.
.PP
These control characters are traditionally used in Fortran programs.
.PP
The first character of every line is removed from input, and the following
actions are performed:
.PP
If the caracter removed is:
.TP 8
.B <space>
the rest of the line is output without change.
.TP 8
.B 0
a
.B <newline>
is output, then the rest of the input line.
.TP 8
.BR 1
a
.B <form feed>
is output, then the rest of the input line.
.TP 8
.B +
An attempt is made to overstrike the previous line with the one following
the character; this can be used for emphasis.
.TP 8
.B <any other character>
The same actions as for a
.B <space>
are performed; this is an extension of the POSIX.2 standard.
.PP
The overstriking generated may be removed by piping the output through
.BR "col -b".
.PP
Tabs in the input file are expanded to a width of eight.
.SH BUGS
It is not be possible to use the
.B asa
filter for interactive use in Fortran programs, because of the semantics of
the
.B +
character, and because I/O is buffered.
.SH "CONFORMS TO"
POSIX.2
.SH "SEE ALSO"
f2c(1), f77(1), col(1)


ld/NEWS
-*- text -*-

Changes in version 2.9:

* Added SQUAD to the linker script language.

Changes in version 2.8:

* Linker scripts may now contain shell wildcard characters for file and section
  names.

* The linker now supports symbol versions in ELF.

* The NOCROSSREFS command was added to the linker script language.

* The LOADADDR expression was added to the linker script language.

* MAX and MIN functions were added to the linker script language.

* The OVERLAY construct was added to the linker script language.

* New option --warn-section-align to warn when the address of an output section
  changes due to alignment of an input section.

* New options --filter/-F and --auxiliary/-f.

Changes in version 2.7:

* New option --cref to print out a cross reference table.

* New option --wrap SYMBOL.

* New option --no-whole-archive, to turn off the effect of --whole-archive.

* Input sections assigned to the output section /DISCARD/ in the linker script
  are not included in the output file.

* The SunOS and ELF linkers now merge stabs debugging information which uses
  the N_BINCL and N_EINCL stab types.  This reduces the amount of debugging
  information generated.

Changes in version 2.6:

* When an ELF section name is representable as a C identifier (this is not true
of most ELF section names), the linker will automatically define symbols
__start_SECNAME and __stop_SECNAME, where SECNAME is the section name, at the
beginning and the end of the section.  This is used by glibc.

* When an ELF section named .gnu.warning is encountered in an input file, the
contents of the section are displayed as an error message, and the section is
not copied into the output file.  This is used by glibc.

* When an ELF section named .gnu.warning.SYMBOL is encountered in an input
file, and the symbol SYMBOL is referenced by some object file, the contents of
the section are displayed as an error message.  The section is not copied into
the output file, unless doing a relocateable or shared link.  This is used by
glibc.

* New options -split-by-reloc and -split-by-file.

* The linker now supports linking PIC compiled code on SPARC SunOS.  It can
also create SPARC SunOS shared libraries, and, like the native SunOS linker,
will do so whenever there is an undefined symbol in the link and neither the -e
nor the -r option was used.

* The -rpath option may be used on SunOS to set the list of directories to be
searched at run time.  This overrides the default of building the list from the
-L options.

* The COFF linker now combines debugging information for structs, unions, and
enums, so that even if the same type is defined in multiple input files it will
only be defined once in the output file.  The --traditional-format switch will
prevent this optimization.

Changes in version 2.5:

* The linker now supports linking against SunOS shared libraries.  It still can
not link SunOS PIC (Position Independent Code) files, so it can not be used to
generate shared libaries.

* The linker now supports linking against ELF shared libraries for the i386
(UnixWare) and SPARC (Solaris).  It can also link ELF PIC files, and can be
used to generate shared libraries.  Shared library generation is not well
tested; please report any problems encountered.  The linker is now enabled for
Solaris again.

* Eric Youngdale has contributed Linux support code, including linking against
Linux a.out shared libraries.  The linker produces Linux QMAGIC binaries.

* The ELF backend has been converted to the new linker code.  To use the new
ELF linker, each particular target requires a relocation function.  So far,
this function has been written for i386 (UnixWare), SPARC (Solaris) MIPS (Irix
5), and HPPA ELF targets.

* The -( (--start-group) and -) (--end-group) options have been added to
support searching a group of archives as though they were a single archive.
This can also be used in a linker script, as GROUP ( files ).

* When a file is named on the command line, and the linker does not recognize
it as an object file, the linker will now treat the file as a linker script
file.  A linker script named in this way augments, but does not replace, the
default linker script.

* The -warn-once option was added.  It causes the linker to only warn once per
undefined symbol, rather than once per reference.

* The COFF backend has been converted to the new linker code.  As with ELF, to
use the new linker, each particular target requires a relocation function.  So
far, this function has been written for the i386, m68k, a29k and SH targets.

* The -V flag was made a synonym for -v, for SVR4 compatibility.  The old -V
behaviour is available via --verbose.

Changes in version 2.4:

* New linker code, by Steve Chamberlain and Ian Taylor.  For a.out and ecoff
  formats (so far), this should result in considerable savings in time
  and memory used while linking; slightly poorer performance than
  before for formats not converted yet.

* Command-line parsing is no longer done with flex.  This means
  oddball characters in filenames won't get treated as argument
  separators.

* HP-PA ELF support, by Jeff Law.  (No SOM support yet.)

* Mach i386 support, by David Mackenzie.

* Irix 4 shared libraries are now supported (Irix 5 uses ELF, and ELF shared
  libraries are not yet supported).

* COFF shared libraries (as on SCO) should work as well.

* The linker is disabled for Solaris.  (Actually, it was in 2.3 also, I just
  forgot to note it.)  Some of their C library routines don't work when
  statically linked, and the GNU linker doesn't support dynamic linking yet.

Changes in version 2.3:

* Weak symbols are now supported.

* ELF support has been added.  The linker has been bootstrapped on
  UnixWare and Solaris.

* Alpha OSF/1 support has been added (non dynamic linking only).

Changes in version 2.2:


asa/asa.c
/* 
 *  asa.c - interpret ASA carriage control characters
 *  Copyright (C) 1994 Thomas Koenig
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#define _POSIX_SOURCE 1

/* System Headers */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

/* Macros */

#define LINESIZE 135
#define NUMLINES 5
#define MAX(a,b) ((a) > (b) ? (a) : (b))

/* Structures and unions */

typedef struct
{
    char *chr;
    size_t len;
} str;

/* File scope variables */

static char rcsid[] = "$Id: asa.c,v 1.1 1998/03/07 00:15:43 SOFTWAY_COM+donn Exp $";
static str *line_buffer = (str *) 0;
static size_t line_num = 0;
static size_t linebuf_size;

/* Function declarations */

static void *mymalloc(size_t);
static size_t readline(FILE *fp, char **a);
static void form_feed(void);
static void add_line(str);
static void flush(void);
static void copy_file(FILE *fp);
  
/* Local functions */

static void*
mymalloc(size_t n)
{
    void *p;
    if ((p=malloc(n)))
	return p;
    else
    {
	fprintf(stderr,"Virtual memory exhausted\n");
	exit(EXIT_FAILURE);
    }
}

static void*
myrealloc(void *p, size_t n)
{
    if ((p=realloc(p,n)))
	return p;
    else
    {
	fprintf(stderr,"Virtual memory exhausted\n");
	exit(EXIT_FAILURE);
    }
}
static void
form_feed()
{
    putchar('\f');
}

static void
new_line()
{
    putchar('\n');
}

static void
add_line(str a)
{
    if (line_num >= linebuf_size)
    {
	linebuf_size += NUMLINES;
	line_buffer = myrealloc(line_buffer, linebuf_size*sizeof(str *));
    }
    line_buffer[line_num]=a;
    line_num++;
}

static void
flush()
{
    size_t i,j;
    size_t max_len;

    if (line_num == 0)
	return;
    if (line_num == 1)
    {
	printf("%s\n", line_buffer[0].chr +1);
	line_num = 0;
	return;
    }
    max_len = 0;
    for (j=0; j< line_num; j++)
	max_len = MAX(max_len, line_buffer[j].len);

    for (i=1; i<= max_len;  i++)
    {
	int printed = 0;

	for (j=0; j< line_num; j++)
	{
	    if (line_buffer[j].len > i)
	    {
		int ch = line_buffer[j].chr[i];

		if (ch != ' ') {
	            if (printed)
			putchar('\b');
		    putchar(ch);
		    printed = 1;
		}
	    }
	}
	if (!printed)
	    putchar(' ');
    }
    for (j=0; j<line_num; j++)
	free(line_buffer[j].chr);

    line_num = 0;
    putchar('\n');
}

static size_t
readline(FILE *fp, char **ret)
{
    static char *buffer=(char *) 0;
    char *ret_buff;
    static int bufsize = LINESIZE;
    int ch;
    size_t len=0;
    int inc;
    int i;

    if (buffer == (char*) 0)
	buffer = (char *) mymalloc(LINESIZE);

    while(1)
    {
        ch = fgetc(fp);
	if (ch == EOF)
	    break;
	if (ch == '\t')
	{
	    ch = ' ';
	    inc = 8 - (len % 8);
	}
	else
	    inc = 1;

	if (len + inc > bufsize-2)
	{
	    bufsize += LINESIZE;
	    buffer = myrealloc(buffer, bufsize);
	}
	for (i=0; i<inc; i++)
	    buffer[len+i] = ch;
	len += inc;
	if (ch == '\n')
	    break;
    }
    buffer[len]='\0';
    ret_buff = mymalloc(len+1);
    strcpy(ret_buff, buffer);
    *ret = ret_buff;
    return len;
}

static void
copy_file(FILE *fp)
{
    str line;
    static first_line = 1;

    while ((line.len=readline(fp, &(line.chr))))
    {
	if (line.chr[line.len-1] == '\n')
	{
	   line.chr[line.len-1] = '\0';
	   line.len--;
	}

	switch (line.chr[0])
	{
	case '+':
	    add_line(line);
	    break;

	case '0':
	    flush();
	    new_line();
	    add_line(line);
	    break;

	case '1':
	    flush();
	    if (! first_line)
		form_feed();
	    add_line(line);
	    break;

	case ' ':
	    flush();
	    add_line(line);
	    break;

	default:
	    flush();
	    add_line(line);
	    break;
	}
        first_line = 0;
    }
}

/* Global functions */

int
main(int argc, char **argv)
{

    line_buffer = (str *) mymalloc(NUMLINES*sizeof(str*));
    linebuf_size = NUMLINES;

    if (argc == 1)
	copy_file(stdin);
    else
    {
	FILE *fp;
	char *fname;

	while(--argc > 0)
	{
	    fname = *++argv;
	    if ((fp = fopen(fname,"r")) == NULL)
	    {
		fprintf(stderr,"Cannot open %s: %s\n",fname,strerror(errno));
		exit(EXIT_FAILURE);
	    }
	    copy_file(fp);
	    fclose(fp);
	}
    }
    flush();
    exit(EXIT_SUCCESS);
}
